typings/styled-components.d.ts

```diff
@@ -0,0 +1,162 @@
+import * as React from "react";
+import { StatelessComponent } from "react";
+
+type ConstrainedProps<C, P> = C & ({ defaultProps?: P } | { new(props?: P, context?: any): any });
+type StyledProps<P> = P & { theme: any };
+
+interface StyledFunction<T, P> {
+  (strs: TemplateStringsArray, ...fns: Array<(props: P) => string>): T;
+}
+
+interface StyledInterface {
+  <C extends React.ComponentClass<P>, P, ThemeInterface>(component: ConstrainedProps<C, StyledProps<P>>): StyledFunction<C, P>;
+  <C extends React.StatelessComponent<P>, P, ThemeInterface>(component: ConstrainedProps<C, P>): StyledFunction<C, P>;
```

Comment:
should `component` be of type `ConstrainedProps<C, StyledProps<P>>`? that would mean it could use `theme` prop.
Also, it should return component with `StyledProps<P>` because resulting component allows to pass `theme` as a prop.